app/services/school_data_importer.rb

```diff
@@ -46,77 +41,119 @@ def school_data_files
-      logger.info "LA name change in school import. Old name: #{local_authority_district.name}, New name: #{row_lad_name}"
+      school.save!
     end
 
-    local_authority_district.name = row_lad_name
-    local_authority_district.save!
-    local_authority_district
-  end
+    def local_authority
+      return @local_authority if @local_authority
+
+      code = data.fetch("LA (code)")
+      @la_cache.fetch(code) do
+        local_authority = LocalAuthority.find_or_initialize_by(code: code)
+        new_name = data.fetch("LA (name)")
 
-  def link_school_to_local_authority(school, local_authority)
-    school_local_authority = SchoolLocalAuthority.find_or_initialize_by(school: school, local_authority: local_authority)
+        if local_authority.name != new_name
+          if local_authority.persisted?
+            logger.info "LA name changed in school import. Old name: #{local_authority.name}, New name: #{new_name}"
+          end
+          local_authority.name = new_name
+          local_authority.save!
+        end
 
-    if school_local_authority.new_record?
-      school_local_authority.start_year = start_year
-      SchoolLocalAuthority.latest.find_by(school: school)&.update!(end_year: start_year)
+        @local_authority = @la_cache[code] = local_authority
+      end
     end
 
-    school_local_authority.save!
-  end
+    def local_district_authority
+      return @local_district_authority if @local_district_authority
 
-  def link_school_to_lad(school, lad)
-    school_lad = SchoolLocalAuthorityDistrict.find_or_initialize_by(
-      school: school,
-      local_authority_district: lad,
-    )
+      code = data.fetch("DistrictAdministrative (code)")
 
-    if school_lad.new_record?
-      school_lad.start_year = start_year
-      SchoolLocalAuthorityDistrict.latest.find_by(school: school)&.update!(end_year: start_year)
+      @lad_cache.fetch(code) do
+        local_authority_district = LocalAuthorityDistrict.find_or_initialize_by(code: code)
+        new_name = data.fetch("DistrictAdministrative (name)")
+
+        if local_authority_district.name != new_name
+          if local_authority_district.persisted?
+            logger.info "LAD name changed in school import. Old name: #{local_authority_district.name}, New name: #{new_name}"
+          end
+          local_authority_district.name = new_name
+          local_authority_district.save!
+        end
+
+        @local_district_authority = @lad_cache[code] = local_authority_district
+      end
     end
 
-    school_lad.save!
+    def link_school_to_local_authority
+      SchoolLocalAuthority.create!(
+        school: school,
+        local_authority: local_authority,
+        start_year: start_year,
+      )
+    end
+
+    def link_school_to_local_district_authority
+      SchoolLocalAuthorityDistrict.create!(
+        school: school,
+        local_authority_district: local_district_authority,
+        start_year: start_year,
+      )
+    end
```

Comment:
Do we always want to `create!`? I know we only do it for new schools, but maybe `find_or_create_by!` is safer?

Reply:
I am not certain about this. At the moment this executes only for new schools, so there should never be any `SchoolLocalAuthority` record for that school. I agree it is slightly safer in case of future changes to this service, but for now this would only add an extra db query. Additionally, if we assume that there might be a record already, then we need to handle it better in case the district changes (which is the main reason we only do this for new schools at the moment - to prevent LA changes to propagate to our db until we know how to handle it better)

Reply:
Fine. I imagine we already have a test which updates a school that has the same LA, right? So if we end up changing the service, and forget about this line, that test would break?

Reply:
Yup, it's here: https://github.com/DFE-Digital/early-careers-framework/pull/540/files#diff-5681fb24e488348c6c99cddc5acd1f5461796e35d49eea62178959f9d0eed6aeR69

Reply:
This doesn't look right to me. One LA has many schools

Reply:
It would be good to get a unique index on this combination I think